lib/hyperreduction/S_OPT.C

```diff
@@ -0,0 +1,619 @@
+            for (int j = 0; j < i - row_offset[myid] && j < row_offset[myid +  1]; j++) {
+                QRmgr.A->mdata[j * QRmgr.A->mm + i] = 0;
+            }
+            if (i < row_offset[myid + 1]) {
+                QRmgr.A->mdata[(i - row_offset[myid]) * QRmgr.A->mm + i] = 1;
+            }
+        }
+
+        // Obtain Q
+        lqcompute(&QRmgr);
+        Matrix* qr_factorized_basis = new Matrix(row_offset[myid + 1] - row_offset[myid],
+                num_basis_vectors, f_basis->distributed());
+        for (int rank = 0; rank < num_procs; ++rank) {
+            gather_block(&qr_factorized_basis->item(0, 0), QRmgr.A,
+                         1, row_offset[rank] + 1,
+                         num_basis_vectors, row_offset[rank + 1] - row_offset[rank],
+                         rank);
+        }
+        delete f_basis_truncated;
+        Vo = qr_factorized_basis;
+
+        free(QRmgr.tau);
+        free(QRmgr.ipiv);
+        free_matrix_data(QRmgr.A);
+    }
+    else
+    {
+        Vo = f_basis_truncated;
+    }
+
+    int num_samples_obtained = 0;
+
+    // Scratch space used throughout the algorithm.
+    double* c = new double [num_basis_vectors];
+
+    vector<set<int> > proc_sampled_f_row(num_procs);
+    vector<map<int, int> > proc_f_row_to_tmp_fs_row(num_procs);
+    int num_f_basis_cols = f_basis_sampled_inv.numColumns();
+    Matrix V1(num_samples,
+              num_basis_vectors,
+              false);
+
+    RowInfo f_bv_max_local, f_bv_max_global;
+    f_bv_max_local.row_val = -1.0;
+    f_bv_max_local.proc = myid;
+    for (int i = 0; i < num_rows; ++i) {
+        double f_bv_val = fabs(Vo->item(i, 0));
+        if (f_bv_val > f_bv_max_local.row_val) {
+            f_bv_max_local.row_val = f_bv_val;
+            f_bv_max_local.row = i;
+        }
+    }
+    MPI_Allreduce(&f_bv_max_local, &f_bv_max_global, 1,
+                  MaxRowType, RowInfoOp, MPI_COMM_WORLD);
+    // Now get the first sampled row of the basis of the rhs->
+    if (f_bv_max_global.proc == myid) {
+        for (int j = 0; j < num_basis_vectors; ++j) {
+            c[j] = Vo->item(f_bv_max_global.row, j);
+        }
+    }
+    MPI_Bcast(c, num_basis_vectors, MPI_DOUBLE,
+              f_bv_max_global.proc, MPI_COMM_WORLD);
+    // Now add the first sampled row of the basis of the RHS to tmp_fs.
+    for (int j = 0; j < num_basis_vectors; ++j) {
+        V1.item(0, j) = c[j];
+    }
+    proc_sampled_f_row[f_bv_max_global.proc].insert(f_bv_max_global.row);
+    proc_f_row_to_tmp_fs_row[f_bv_max_global.proc][f_bv_max_global.row] = 0;
+    num_samples_obtained++;
+
+    // Square Vo.
+    Matrix* nVo = Vo->elementwise_square();
+
+    Vector* A = new Vector(num_rows, f_basis->distributed());
+    Vector* noM = new Vector(num_rows, f_basis->distributed());
+
+    for (int i = 2; i <= num_samples; i++)
+    {
+        if (i <= num_basis_vectors)
+        {
+            Matrix A0(i - 1, i - 1, false);
```

Comment:
Instead of allocating these matrices for every iteration, it would be more efficient to declare `Matrix` objects of the maximum size to be used, and then use `setSize` on each iteration to use the same allocated memory.